packages/serverpod/lib/src/server/serverpod.dart

```
@@ -881,3 +969,50 @@ extension ServerpodInternalMethods on Serverpod {
   /// Logging is turned off.
   Session get internalSession => _internalSession;
 }
+
+/// Container for a list of [DiagnosticEventHandler]s
+/// that will run concurrently with each other
+/// and asynchronously with the caller.
+class _EventHandlers implements DiagnosticEventHandler {
+  final List<DiagnosticEventHandler> handlers;
+
+  /// If set, this timeout is applied to each event handler invocation.
+  final Duration? timeout;
+
+  /// Creates a new [_EventHandlers] with the specified list of handlers.
+  const _EventHandlers(
+    this.handlers, {
+    // ignore: unused_element
+    this.timeout = const Duration(seconds: 30),
+  });
+
+  @override
+  void handleEvent(
+    DiagnosticEvent event,
+    OriginSpace space, {
+    required DiagnosticEventContext context,
+  }) {
+    var futures = handlers.map((handler) => Future(
+          () => handler.handleEvent(event, space, context: context),
+        ));
+
+    var to = timeout;
+    if (to != null) {
+      futures = futures.map((future) => future.timeout(to));
+    }
+
+    futures.wait.onError((ParallelWaitError e, stackTrace) {
```

Question: Is this on error triggered for all of them or only for the first one? I remember we stumbled a bit on that once before.

---

It is triggered if at least one throws. All are still run. From the `.wait` documentation:

If any future completes with an error, the returned future completes with a [ParallelWaitError].

---

Yes, but is it repeatedly triggered if multiple of the futures throws?

My concern is that we might silence errors and only print a warning for the first Future that throws.

The documentation for `wait` states:

> If any future completes with an error, then the returned future completes with that error. If further futures also complete with errors, those errors are discarded.

https://api.flutter.dev/flutter/dart-async/Future/wait.html

That is not desired behavior right?

---

That's the wrong wait, it's this one:
https://api.flutter.dev/flutter/dart-async/FutureIterable/wait.html

Yes all the futures are completed.

---

Aah, my mistake! thanks for the clarification!